### ISSUE TITLE
keras future: use nn for representational learning early in ensemble

### ISSUE THREAD

**Issue by ClimbsRocks** [ClimbsRocks/auto_ml#178]
neural nets are fantastic for representational learning (turning raw data inputs into features). gradient boosting is fantastic at turning good features into accurate predictions. 

why not use both?

have a nn that does something like [1, 0.5. 0.1] with an output layer. drop the output layer, and just get the features from the last hidden layer. 

for each row, feed these features into a gradient boosting model (along with all the other features we'd normally use). 

voila- both models are used for what they're best at!

**Comment by ClimbsRocks** [comment]
Thoughts

- We'll want to have the user pass in two different data sets- one for the final model training, and one for training the deep learning network used for feature learning. 
- I really don't want to run two feature transformation pipelines. 
- We have to be really careful with overfitting around feature selection
  - i think we might just have to handle each step of the pipeline surgically
  - so, we'll fit the transformation pipeline on all the data (fl - feature learning- data and final model data), and the transformation pipeline up to the step of feature selection
  - then we'll transform the data for our feature learning model (no feature selection here)
  - Then we'll fit the feature learning model
    - We'll train it on the fl dataset like a standard regression/classification model
    - then we'll remove the final step and just get the learned features from the penultimate step
  - then we'll add this to our pipeline- after dataframe vectorizer, and before feature selection




Thoughts on the feature learning model itself
- this is essentially just going to be a version of train up our final model
- differences
- we might be able to use finalmodelatc, and add in a transform method!
- if transform, then simply add the model's prediction to X
- if predict, do what it currently does
- this allows us to leverage our entire current architecture. we just need to add a transform to finalmodelatc, add a param for 'feature_learning_model', and add in some logic for how we add this to our pipeline (which will be nontrivial, since we need to add it to our dataframevectorizer as well since that's where our feature names are stored which we need for analytics, which will still be necessary here)

the more i think about this, the more i think we need to make a decision about where we do and do not support feature selection throughout this process. 
- the difficulty is that right now i'm envisioning adding in our ensembled model predictions after dataframevectorizer, and before feature_selection. 
- then remember that we smash these two into one (we just use feature selection as a mask for dfv)
- which means things'll get messy if we decide not to include an ensembled prediction due to feature selection
- simplifying assumption: no feature selection at all for the v0 launch. just add a quick if check, and a user warning. 


- we'll want this to be generalizable to other forms of ensembling
- take in a param for 'feature_learning_model' or something, that lets us know whether this is a deep learning network we need to remove the final layer from or not
- take in a param for model name
- take in a param

**Comment by ClimbsRocks** [comment]
finished, with `feature_learning`.
